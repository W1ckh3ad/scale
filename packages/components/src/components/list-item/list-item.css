:host(scale-list-item) {
  display: block;
}

.list-item {
  position: relative;
  font-size: var(--type-size-3, 1rem);
  line-height: var(--type-leading-3, 1.5);
  padding-left: calc(0.25ch + var(--spacing-4, 1rem));
}

.list-item:before {
  left: 0;
  content: '';
  position: absolute;
}

.list-item--unordered:before {
  top: 0.5em;
  border: 1px solid currentColor;
  display: block;
  transform: scale(0.66);
  background: currentColor;
  box-sizing: border-box;
  border-radius: 50%;
  width: 0.5rem;
  height: 0.5rem;
}

.list-item--nested.list-item--unordered:before {
  border: 1px solid currentColor;
  background: transparent;
}

.list-item--ordered {
}

.list-item--ordered:before {
  content: attr(data-index) '.';
  font-size: var(--type-size-3, 1rem);
  font-weight: var(--type-weight-bold, 700);
}

.list-item--nested.list-item--ordered:before {
  font-size: var(--type-size-2, 0.75rem);
  padding-top: calc(var(--type-size-3, 1rem) - var(--type-size-2, 0.75rem));
}

.list-item--no-marker {
  display: inline-flex;
  align-items: center;
  padding-left: 0;
}

.list-item--no-marker:before {
  display: none;
}

.list-item--no-marker > :slotted(*) {
  margin-right: var(--spacing-2, 0.5rem);
}
