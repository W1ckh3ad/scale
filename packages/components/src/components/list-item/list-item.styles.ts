import { JssStyle } from 'jss';

export const styles: JssStyle = {
  '@global': {
    ':host(scale-list-item)': {
      display: 'block',
    },
  },

  'list-item': {
    position: 'relative',
    paddingLeft: ({ spacing }) => `calc(0.25ch + ${spacing[4]})`,

    '&::before': {
      content: '""',
      position: 'absolute',
      left: 0,
    },
  },

  'list-item--nested': {},

  'list-item--unordered': {
    '&::before': {
      boxSizing: 'border-box',
      display: 'block',
      top: '0.5ch',
      width: ({ spacing }) => spacing[2],
      height: ({ spacing }) => spacing[2],
      borderRadius: '50%',
      transform: 'scale(0.66)',
      background: 'currentColor',
    },

    '$list-item--nested&::before': {
      background: 'transparent',
      border: '1px solid currentColor',
    },
  },

  'list-item--ordered': {
    '&::before': {
      content: 'attr(data-index) "."',
      fontSize: ({ type }) => type.size_3,
      fontWeight: ({ type }) => type.weight_bold,
    },

    '$list-item--nested&::before': {
      fontSize: ({ type }) => type.size_2,
    },
  },

  'list-item--no-marker': {
    display: 'inline-flex',
    alignItems: 'center',
    paddingLeft: 0,

    '&::before': {
      display: 'none',
    },

    '& > ::slotted(*)': {
      marginRight: ({ spacing }) => spacing[2],
    },
  },

  'list-item__nested-list': {},
};
