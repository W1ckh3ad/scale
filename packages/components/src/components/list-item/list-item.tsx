import { Component, h, Prop, Host, Element, State } from '@stencil/core';
import classNames from 'classnames';

/**
 * @todo styles for custom icon (no-marker prop?)
 * @see https://github.com/carbon-design-system/carbon-web-components/tree/master/src/components/list
 */

@Component({
  tag: 'scale-list-item',
  styleUrl: './list-item.css',
  shadow: true,
})
export class ListItem {
  @Element() el: HTMLElement;

  /** Whether this is a child of an ordered scale-list, gets set automatically by its parent */
  @Prop() ordered?: boolean = false;
  /** Index number, useful only for styling the `ordered` type */
  @Prop({ mutable: true }) index?: number;
  /** If `false`, no marker or left padding will be visible */
  @Prop() marker: boolean = true;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  @State() hasNestedChild: boolean = false;
  @State() isNested: boolean = false;

  componentWillLoad() {
    this.isNested = this.isNestedCheck();
  }

  connectedCallback() {
    if (!this.el.hasAttribute('role')) {
      this.el.setAttribute('role', 'listitem');
    }
  }

  handleSlotChange = ({ target }: Event) => {
    this.hasNestedChild =
      (target as HTMLSlotElement).assignedNodes().length > 0;
    this.isNested = this.isNestedCheck();
  };

  isNestedCheck = (): boolean => {
    return this.el.closest('scale-list[slot="nested"]') != null;
  };

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}

        <div class={this.getCssClassMap()} data-index={this.index}>
          <slot />
          <div class={`list-item__nested-list`} hidden={!this.hasNestedChild}>
            <slot name="nested" onSlotchange={this.handleSlotChange}></slot>
          </div>
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    const orderType = this.ordered ? 'ordered' : 'unordered';

    return classNames(
      'list-item',
      this.isNested && 'list-item--nested',
      `list-item--${orderType}`,
      !this.marker && 'list-item--no-marker'
    );
  }
}
