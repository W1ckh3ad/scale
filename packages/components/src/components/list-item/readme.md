# scale-list-item



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                    | Type  | Default     |
| -------- | --------- | ------------------------------ | ----- | ----------- |
| `styles` | `styles`  | (optional) Injected jss styles | `any` | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
