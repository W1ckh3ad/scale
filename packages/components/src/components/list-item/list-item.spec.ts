import { newSpecPage } from '@stencil/core/testing';
import { ListItem } from './list-item';

describe('ListItem', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [ListItem],
      html: `<scale-list-item>default</scale-list-item>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
