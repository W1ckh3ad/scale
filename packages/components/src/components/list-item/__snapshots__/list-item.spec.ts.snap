// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should match snapshot 1`] = `
<scale-list-item role="listitem">
  <mock:shadow-root>
    <style>
      :host {
  display: block;
}
.list-item-0-0-1 {}
.list-item--nested-0-0-2 {
  color: tomato;
}
.list-item__nested-list-0-0-3 {
  border: 1px solid gold;
}
    </style>
    <div class="list-item-0-0-1">
      <slot></slot>
      <div class="list-item__nested-list-0-0-3" hidden="">
        <slot name="nested"></slot>
      </div>
    </div>
  </mock:shadow-root>
  default
</scale-list-item>
`;
